R/buffer.R

```diff
@@ -0,0 +1,84 @@
+#' Apply an inclusion radius and exclusion buffer to indices
+#'
+#' @param data An object of class `sf` or `sfc`.
+#' @param indices List of indices in each fold generated by `split_unnamed`.
+#' @param radius Numeric: points within this distance of the initially-selected
+#' test points will be assigned to the assessment set.
+#' @param buffer Numeric: points within this distance of the test set
+#' (after `radius` is applied)
+#' will be assigned to neither the analysis or assessment set.
+#'
+#' @keywords internal
+buffer_indices <- function(data, indices, radius, buffer) {
+
+  if (identical(sf::st_crs(data), sf::NA_crs_)) {
+    rlang::abort(
+      c("`buffer` and `radius` require `data` to have a non-NA coordinate reference system",
+        "i" = "Set the CRS for your data using `sf::st_set_crs()`"
+      )
+    )
+  }
+
+  if (sf::st_is_longlat(data) && !sf::sf_use_s2()) {
+    rlang::abort(
+      c(
+        "`buffer` and `radius` can only be used with geographic coordinates when using the s2 geometry library",
+        "i" = "Reproject your data into a projected coordinate reference system using `sf::st_transform()`",
+        "i" = "Or install the `s2` package and enable it using `sf::sf_use_s2(TRUE)`"
+      )
+    )
+  }
+
+  n <- nrow(data)
+
+  if (is.null(radius)) radius <- 0
+  if (is.null(buffer)) buffer <- 0
+
+  if (radius) {
+    indices <- row_ids_within_dist(data, indices, radius)
+  }
+
+  # `buffer_indices` are _always_ needed
+  # so initialize to integer(0) before checking if we need to find them:
+  buffer_indices <- lapply(seq_along(indices), function(x) integer(0))
```

Comment:
Let me know how you feel about my change re: this. I left the `if (radius)` check, because I found myself having to write really circuitous code to avoid that check and still combine `indices` with the new indices (and to do so in a way that concatenates the two lists, instead of sorting them in).